### ISSUE TITLE
Escape quotes (' or ") in lists for uBO

### ISSUE THREAD

**Issue by Yuki2718** [AdguardTeam/FiltersCompiler#156]
Due to new cosmetic filter parser, uBO requires quotes in arguments to be escapsed: https://github.com/gorhill/uBlock/commit/a71b71e4c8a2037fc68970bc8912a76732edaade

An example that needs the fix is `thetimestribune.com#?##tncms-region-option-rail-top-one > .tncms-block:has(> .block-title:contains(This Week's Circulars))` in Base filter.

**Comment by Yuki2718** [comment]
Probably only `:contains` will be affected in AG lists.

**Comment by Alex-302** [comment]
https://github.com/AdguardTeam/AdguardFilters/pull/134630#issuecomment-1313741190

**Comment by Yuki2718** [comment]
@Alex-302 It's totally different. The PR was about an uBO issue that will eventually be fixed but the problem here is permanent change of uBO.

**Comment by krystian3w** [comment]
https://github.com/AdguardTeam/ExtendedCss/issues/151

**Comment by scripthunter7** [comment]
@Yuki2718 FYI: We have created the ECSSTree library, which is fully backwards compatible with basic CSSTree, but thanks to re-tokenization it can parse the edge case you mentioned and many other cases as well. You can find it at the following link: https://github.com/AdguardTeam/ecsstree

**Comment by scripthunter7** [comment]
@Yuki2718 FYI: We have created the ECSSTree library, which is fully backwards compatible with basic CSSTree, but thanks to re-tokenization it can parse the edge case you mentioned and many other cases as well. You can find it at the following link: https://github.com/AdguardTeam/ecsstree

**Comment by krystian3w** [comment]
Also a little outdated is escape method, now works only:

> ```adb
> thetimestribune.com#?##tncms-region-option-rail-top-one > .tncms-block:has(> .block-title:contains("This Week's Circulars"))
> ```
> 

to avoid regex syntax for that as long uBo consider switch to **E**CSSTree.

**Comment by scripthunter7** [comment]
> Same protection needed **orphaned** `( )`

If the balancing is not correct, it is impossible to clearly determine the actual end of the pseudo-class. For example:
- `:contains((aa)`
- `:contains(aa))`

In the above cases, parenthesis must be escaped as follows:
- `:contains(\(aa)`
- `:contains(aa\))`

But the following case is fine, since balancing is fine:
- `:contains((aa))`

The following cases are currently problematic in ECSSTree:
- `:contains('(aa')`
- `:contains('aa)')`

This is because ECSSTree currently ONLY parses the argument as raw (and not takes string into account). If we introduce the string, the following case must also be taken into account:
- from `:contains('aa"aa')`, `'aa"aa'` is parsed as string node by CSSTree, but when generating, CSSTree always uses double quotes, so it generate this selector as `:contains("aa\"aa")` ([reference](https://github.com/csstree/csstree/issues/211))
- `:contains('aaa)`
- `:contains("aaa)`

I will think about this before the v1.0.8 release

**Comment by scripthunter7** [comment]
> Same protection needed **orphaned** `( )`

See https://github.com/AdguardTeam/ecsstree/issues/11

**Comment by krystian3w** [comment]
I'm curious if ECSSTree is used to compile filters in the "/ublock/" directory - if that is, the AdGuard filter problem fixed outside of [`FiltersCompiler`](https://github.com/AdguardTeam/FiltersCompiler/) / [`coreLibs`](https://github.com/AdguardTeam/coreLibs/) (it is possible that here almost never) / [`tsurlfilter`](https://github.com/AdguardTeam/tsurlfilter/) / [`FiltersRegistry`](https://github.com/AdguardTeam/FiltersRegistry).

Quickly I can no longer find any filters what would have `'` `"` in contains (simple regex can fail for contains with orphaned bracket).

*(I did the same myself with one detection of incompatible filter - for me it was enough to check less characters)*

The one from the demo I think someone has already rewritten/removed by hand (But can be cumbersome for new cases).

**Comment by scripthunter7** [comment]
> AdGuard filter problem

I don't think this is a "problem" of AdGuard filters only, but a more general problem, since this syntax is used in all adblockers, as far as I know. It's simply that CSSTree's tokenizer doesn't support this syntax (for obvious reasons). CSSTree is an excellent library for various purposes, I have been experimenting with it since June 2022, that's why ECSSTree was created to solve these annoying edge cases. I tested ECSSTree on the entire AG Filters Registry and the entire uAssets, so I handled several edge cases. We will use ECSSTree in a static analyzer tool, which should parse all types of adblock rules, but I think it can be used for any other purpose as long as it helps working with Extended CSS language elements.
